### ISSUE TITLE
Model Interfaces

### ISSUE THREAD

**Issue by Leland-Takamine** [yahoo/squidb#22]
Is there any way to generate a model that implements an interface?

**Comment by sbosley** [comment]
Not at the moment, but I'm willing to consider it! Should be pretty straightforward to add to the code generator.

In the meantime, you can use [ModelMethod](https://github.com/yahoo/squidb/wiki/Advanced-model-spec-options#modelmethod) to add methods to the generated models until we can figure out the best way to add the ability to explicitly declare "implements SomeInterface".

**Comment by sbosley** [comment]
Can you share a bit more about your use case for this feature? I want to make sure we design it in a way that will work for what you need and that some other (perhaps hidden) feature doesn't already solve the need.

**Comment by Leland-Takamine** [comment]
Thanks for the response @sbosley! After more thought I don't think my specific use case would have been addressed by model interfaces, but I'll note here what I was trying to do anyway.

I have ~10 tables that share a column (parentID). I need a way to clear all ~10 of these tables given a parentID. It's possible to hardcode a "clear(parentID)" method for each table but it seems like there should be a better way to do this.

If tables have columns in common, I'd like a way to write a generic method that executes SQL on one of these tables without having to hard code exactly which table we're dealing with.

Please let me know if you'd like any more details.

**Comment by sbosley** [comment]
Ahh, I see. I think there's one way in which interfaces would help you a little bit, but I think what you could do in the interim is define a raw Field, which is basically just a wrapper around a column name not associated with a specific table. Field objects aren't very useful in most cases, but you can use them to construct Criterions, so I think they would work well for this use case:

```java
// Assuming parentID is a string; the type parameter isn't super important here
private static final Field<String> PARENT_ID = Field.field("parentID");

public void clearTableByParentId(Class<? extends TableModel> modelClass, String parentId) {
    databaseDao.deleteWhere(modelClass, PARENT_ID.eq(parentId));
}
```

This would work as long as all the tables/model classes you ever passed to it declared a column named "parentID". But in an ideal world that method would ONLY accept model classes who have the parent id column, right? So it would be useful if those classes implemented a common interface, because then you could do this and add some nice compile time safety:
```java
public <T extends TableModel & ITableWithParent> void clearTableByParentId(
    Class<T> modelClass, String parentId) {
    databaseDao.deleteWhere(modelClass, PARENT_ID.eq(parentId));
}
```

The interface ITableWithParent itself would be empty (no actual methods), but you could at least use it to declare at compile time "this model is safe to use when parentID is required". I can still see how that would be a legitimate use case, so if that sounds like it would be useful to you I'll keep the issue open.

Hopefully that answers your question! I'm definitely intrigued by the use case so let me know if any of my suggestions are useful to you, or if you have something even more specific in mind that would work better for what you need.

**Comment by Leland-Takamine** [comment]
Wow thank you! This is a perfectly workable solution for now, and the interface solution would be a great improvement as well. Really appreciate the suggestions and looking forward to working with this library a bit more.

**Comment by MFlisar** [comment]
I would need the interfaces as well.. My use case is, that I have to data soruces, one from squidb, one from phone, that offer very similar data. I want to use them equally in my adapters and source code, so I would like to define an interface for those two classes...

There's no hidden way to this, is there?

**Comment by sbosley** [comment]
There's no hidden way to do this currently, but I do have an unpushed branch in progress to add the functionality. It should be ready soon!

**Comment by MFlisar** [comment]
Sounds great. Currently I've written a wrapper class for my use case, using interfaces would make everything much cleaner

**Comment by sbosley** [comment]
Hey all, just merged a PR to add this feature. We'll add some more thorough docs to the wiki later today, but the gist of it is to add an annotation declaring the interfaces to your model spec, then use `@ModelMethod` to implement the required methods:

```java
@TableModelSpec(...)
@Implements(interfaceClasses = Runnable.class)
public class ModelSpec {
    ...
    
    @ModelMethod
    public static void run(Model instance) {
        // Implement run() for the instance here
    }
}
```

The feature is merged into the master branch and will be available in the pre-built artifacts when we launch 1.0.2, hopefully later this week.

**Comment by sbosley** [comment]
This was released in our launch of 1.1.0 today. Upgrade your dependencies and check it out!

**Comment by MFlisar** [comment]
works great, exactly what I wanted. thanks
